templates/tree-widget/treeview-example-model.ts

```diff
@@ -0,0 +1,202 @@
+}
+export namespace ExampleTreeLeaf {
+    export function is(candidate: object): candidate is ExampleTreeLeaf {
+        return TreeNode.is(candidate) && 'type' in candidate && candidate.type === 'leaf';
+    }
+}
+
+/**
+ * Example data to initialize the "business model"
+ */
+const EXAMPLE_DATA: Item[] = [{
+    name: 'Fruits',
+    children: [
+        {
+            name: 'Apples',
+            children: [
+                { name: 'Golden Delicious', quantity: 4 },
+                { name: 'Gala', quantity: 3 },
+            ]
+        },
+        {
+            name: 'Oranges',
+            children: [
+                { name: 'Clementine', quantity: 2 },
+                { name: 'Navel', quantity: 5 },
+            ]
+        }
+    ]
+},
+{
+    name: 'Vegetables',
+    children: [
+        { name: 'Carrot', quantity: 10 },
+        { name: 'Zucchini', quantity: 6 },
+        { name: 'Broccoli', quantity: 8 },
+    ]
+}
+];
+
+/**
+ * The Tree Model for the tree.
+ *
+ * This class contains the bridge between business model and tree model and realizes operations on the data.
+ */
+@injectable()
+export class TreeViewExampleModel extends TreeModelImpl {
+
+    /**
+     * Initialize the tree model from the business model
+     */
+    @postConstruct()
+    protected override init(): void {
+        super.init();
+
+        // create the root node
+        const root: CompositeTreeNode = {
+            id: ROOT_NODE_ID,
+            parent: undefined,
+            children: [],
+            visible: false // do not show the root node in the UI
+        };
+
+        // populate the direct children
+        EXAMPLE_DATA.map(item => Item.toTreeNode(item))
+            .forEach(node => CompositeTreeNode.addChild(root, node));
+
+        // set the root node as root of the tree
+        // This will also initialize the ID-node-map in the tree, so this should be called
+        // after populating the children.
+        this.tree.root = root;
+    }
+
+    /**
+     * This is executed when a tree item's checkbox is checked/unchecked.
+     *
+     * For this example, the check state is applied to the business model (backOrdered property).
+     *
+     * @param node the affected node
+     * @param checked the new state of the checkbox
+     */
+    override markAsChecked(node: TreeNode, checked: boolean): void {
+        if (ExampleTreeLeaf.is(node)) {
+            node.data.backOrdered = checked;
+        }
+        super.markAsChecked(node, checked);
+    }
+
+    /**
+     * Logic to add a new child item to the given parent.
+     *
+     * For simplicity, we use a static/constant child, so we don't have to implement UI to ask the user for the name etc.
+     * Note that because of the TreeNode.id initialization to Item.name, this method should only be called once. Otherwise
+     * we end up with multiple tree items with the same ID, which is not desirable.
+     *
+     * So in practice, the id should be calculated in a better way...
```

Comment:
If it is not a lot of work, I'd really like to see that we do proper ID calculation on our data items/nodes as this is a central aspect and this should serve as a template.